cxx/pclean/schema.cc

```diff
@@ -0,0 +1,117 @@
+
+PCleanSchemaHelper::PCleanSchemaHelper(const PCleanSchema& s): schema(s) {
+  compute_class_name_cache();
+  compute_ancestors_cache();
+}
+
+void PCleanSchemaHelper::compute_class_name_cache() {
+  for (size_t i = 0; i < schema.classes.size(); ++i) {
+    class_name_to_index[schema.classes[i].name] = i;
+  }
+}
+
+void PCleanSchemaHelper::compute_ancestors_cache() {
+  for (auto const& c: schema.classes) {
+    if (ancestors.find(c.name) == ancestors.end()) {
+      ancestors[c.name] = compute_ancestors_for(c.name);
+    }
+  }
+}
+
+std::set<std::string> PCleanSchemaHelper::compute_ancestors_for(
+    const std::string& name) {
+  std::set<std::string> ancs;
+  std::set<std::string> parents = get_parent_classes(name);
+  for (auto const& p: parents) {
+    ancs.insert(p);
+    if (ancestors.find(p) == ancestors.end()) {
+      ancestors[p] = compute_ancestors_for(p);
+    }
+    ancs.insert(ancestors[p].begin(), ancestors[p].end());
+  }
+  return ancs;
+}
+
+PCleanClass PCleanSchemaHelper::get_class_by_name(const std::string& name) {
+  return schema.classes[class_name_to_index[name]];
+}
+
+std::set<std::string> PCleanSchemaHelper::get_parent_classes(
+    const std::string& name) {
+  std::set<std::string> parents;
+  PCleanClass c = get_class_by_name(name);
+  for (auto const& v: c.vars) {
+    if (std::holds_alternative<ClassVar>(v.spec)) {
+      parents.insert(std::get<ClassVar>(v.spec).class_name);
+    }
+  }
+  return parents;
+}
+
+std::set<std::string> PCleanSchemaHelper::get_ancestor_classes(
+    const std::string& name) {
+  return ancestors[name];
+}
+
+std::set<std::string> PCleanSchemaHelper::get_source_classes(
+    const std::string& name) {
+  std::set<std::string> sources;
+  for (auto const& c: ancestors[name]) {
+    const std::set<std::string>& parents = get_parent_classes(c);
+    if (parents.empty()) {
+      sources.insert(c);
+    }
+  }
+  return sources;
+}
+
+std::string get_base_relation_name(
+    const PCleanClass& c, std::vector<std::string>& field_path) {
+  assert(field_path.size() == 2);
+  std::string& class_var = field_path[0];
+  std::string& var_name = field_path[1];
+  std::string class_name = "";
+  for (const auto& v : c.vars) {
+    if (v.name == class_var) {
+      class_name = std::get<ClassVar>(v.spec).class_name;
+      break;
+    }
+  }
+  assert(class_name != "");
+  return class_name + ':' + var_name;
+}
+
+T_schema PCleanSchemaHelper::make_hirm_schema() {
+  T_schema tschema;
+  for (auto const& c : schema.classes) {
+    for (auto const& v : c.vars) {
+      std::string rel_name = c.name + ':' + v.name;
+      if (std::holds_alternative<DistributionVar>(v.spec)) {
+        T_clean_relation relation;
+        DistributionVar dv = std::get<DistributionVar>(v.spec);
+        relation.distribution_spec = DistributionSpec(
+            dv.distribution_name, dv.distribution_params);
+        relation.is_observed = false;
+        for (const std::string& sc : get_source_classes(c.name)) {
```

Comment:
I think we read this sentence differently:

> For each class $C$, for each attribute $a \in \mathcal{A}_C$, we introduce a \textit{relation} $r_{C.a}$, with output model $\tau_a$, and input classes $I_{r_{C.a}} = \{\textit{src}(\rho) \mid \rho \in \textit{Paths}(\mathcal{G}(\mathcal{S}), \textit{dst}=C)\}$

If we have a graph of classes like A -> B -> C, I think $I_{r_{C.a}}$ should contain both A and B (i.e. there are 2 paths ending at C, one that starts at A and one that starts at B). You seem to have interpreted it as implying there's a single path starting at A, and $I_{r_{C.a}}$ contains only A. (So according to my read, instead of `get_source_classes` this should be `get_ancestor_classes`.) @alex-lew could you clarify that sentence in the Overleaf?

Also, I'm pretty sure `relation.domains` should contain the class `c.name` itself, in addition to its ancestors/sources.

Reply:
Added c.name to the domains.